caffe2/operators/mean_op.h

```diff
@@ -84,26 +123,39 @@ class MeanGradientOp : public Operator<Context> {
 
   template <typename T>
   bool DoRunWithType() {
-    auto& dY = Input(0);
-    const auto* dY_data = dY.template data<T>();
-    int size = dY.numel();
-
     int num_inputs = OutputSize();
     float scale = 1.0f / num_inputs;
-
-    // dX0 = scale * dY
-
-    auto* dX0 = Output(0, dY.sizes(), at::dtype<T>());
-    math::Scale(
-        size, scale, dY_data, dX0->template mutable_data<T>(), &context_);
-
-    // Copy the rest dX
-    for (int i = 1; i < num_inputs; i++) {
-      auto* cur_dX = Output(i);
-      cur_dX->ResizeLike(dY);
-      cur_dX->CopyFrom(*dX0, true /*async*/);
+    std::vector<int> output_dims;
+    auto& dY = Input(0);
+    const auto* dY_data = dY.template data<T>();
+    std::copy(
+        dY.sizes().cbegin(),
+        dY.sizes().cend(),
+        std::back_inserter(output_dims));
+    for (int i = num_inputs; i >= 1; i--) {
+      std::vector<int> input_dims;
+      const auto& original_input_size = Input(i).sizes();
+      std::copy(
+          original_input_size.cbegin(),
+          original_input_size.cend(),
+          std::back_inserter(input_dims));
+      std::vector<int> input_axes;
+      std::vector<int> output_axes;
+      elementwise_ops_utils::ComputeBinaryBroadcastBackwardAxes(
+          input_dims, output_dims, &input_axes, &output_axes);
+      auto* dX = Output(i - 1, original_input_size, at::dtype<T>());
+      auto* dX_data = dX->template mutable_data<T>();
+      math::ReduceSum(
+          output_dims.size(),
+          output_dims.data(),
+          input_axes.size(),
+          input_axes.data(),
+          T(1),
+          dY_data,
+          dX_data,
+          &context_);
+      math::Scale(dX->numel(), scale, dX_data, dX_data, &context_);
```

Comment:
There is no need to do this scale. Just pass scale to the 5th arg of math::ReduceSum.